File: trunk/x10.runtime.cws/src/x10/runtime/cws/Job.java
package x10.runtime.cws;


import static x10.runtime.cws.ClosureStatus.*;

import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.Future;


public abstract class Job extends Closure implements Future {
	public volatile int result;
	public int resultInt() { return result;}
	public static class JobFrame extends Frame {
		
		volatile public int PC;
		public int x;
		public JobFrame() {
			super();
		}
		public Closure makeClosure() {
			assert false;
			return null;
		}
		public void setOutletOn(final Closure c) {
			if (false && Worker.reporting) {
				System.out.println(Thread.currentThread() + ": " + this + " sets outlet on " + c);
			}
			c.setOutlet(
					new Outlet() {
						
						public void run() {
							JobFrame f = (JobFrame) c.parentFrame();
							int v = c.resultInt();
							if (false && Worker.reporting)
								System.out.println(Thread.currentThread() + " " + this 
										+ " sets x on " + f + " to " + v);
							f.x = v;
						}
						public String toString() { return "OutletInto x from " + c;}
						});
		}
		public String toString() {
			return "JobFrame(#" + hashCode() + " " + x + ", PC=" + PC+")";
		}
	}
	final Pool pool;
	public Job(Pool pool) {
		super(new JobFrame());
		this.pool=pool;
		parent = null;
		joinCount=0;
		status = READY;
		
	}
	public static final int LABEL_1=1, LABEL_2=2, LABEL_3=3;
	@Override
	protected void compute(Worker w, Frame frame) {
		JobFrame f = (JobFrame) frame;
		
		if (f.PC!=LABEL_1 && f.PC!=LABEL_2) {
			f.PC=LABEL_1;
			// spawning
			try {
				int x = spawnTask(w);
				if (w.popFrameCheck(x)) return;
				f.x=x;
			} catch (StealAbort z) {
				
				return;
			}
		}
		if (f.PC < LABEL_2) {
			f.PC=LABEL_2;
			if (sync()) {
				return;
			}
		}
		result=f.x;
		setupReturn();
		return;
	}
	
	abstract public int spawnTask(Worker ws) throws StealAbort;
	public void completed() {
		super.completed();
		//System.out.println("Completed " + this + " result=" + result);
        synchronized(this) {
            notifyAll();
        }
        pool.jobCompleted();
    }
    
    
    public boolean isCancelled() { return false;}
    
    public boolean cancel(boolean b) { return false;}

    /**
     * Return result or throw exception using Future conventions
     */
    Object futureResult() throws ExecutionException {
        
        return resultObject();
    }

    public synchronized Object get() throws InterruptedException,
                                       ExecutionException {
        while (!isDone())
            wait();
        return futureResult();
    }

    public synchronized int getInt() throws InterruptedException, ExecutionException {
    	while (! isDone()) wait();
    	return resultInt();
    }
    public synchronized Object get(long timeout, TimeUnit unit)
        throws InterruptedException, ExecutionException, TimeoutException {
        if (!isDone()) {
            long last = System.nanoTime();
            long nanos = unit.toNanos(timeout);
            while (!isDone()) {
                long now = System.nanoTime();
                nanos -= now - last;
                last = now;
                if (nanos > 0)
                    TimeUnit.NANOSECONDS.timedWait(this, nanos);
                else
                    throw new TimeoutException();    
            }
        }
        return futureResult();
    }
}

File: x10-before-local-classes/tags/CVS2SVN_CONVERSION/x10.lib/src/sched/Job.cc
/*
 * (c) Copyright IBM Corporation 2007
 *
 * $Id: Job.cc,v 1.21 2007-12-26 07:57:34 srkodali Exp $
 * This file is part of X10 Runtime System.
 */

#include <x10/xws/Job.h>
#include <x10/xws/Frame.h>
#include <x10/xws/Lock.h>
#include <x10/xws/Closure.h>
#include <x10/xws/Cache.h>
#include <x10/xws/Worker.h>
#include <x10/xws/Pool.h>
#include <x10/xws/Sys.h>
#include <x10/xassert.h>
#include <pthread.h>
#include <iostream>

using namespace std;
using namespace x10lib_xws;

/*-----------------JobFrame----------------------*/

Closure *JobFrame::makeClosure() {
	assert(0);
	return NULL;
}
void JobFrame::setOutletOn(Closure *c) {
  /*Outlets are deleted by the closures to which they are attached*/
  ResultOutlet *t = new ResultOutlet(c, this);
  assert(t != NULL);
  c->setOutlet(t);
}
		
JobFrame *JobFrame::copy() { return new JobFrame(*this); }

/*-------------------Outlet------------------------*/

#if defined(MEM_DEBUG) && (MEM_DEBUG!=0)
volatile int Outlet::nCons=0;
volatile int Outlet::nDestruct=0;
#endif

/*------------------ResultOutlet-------------------*/

// ResultOutlet::ResultOutlet(): Outlet() { }
ResultOutlet::ResultOutlet(Closure *c, JobFrame *f) 
  : Outlet() {
  closure = c;
  jframe = f;
}
void ResultOutlet::run() {
  //  cerr<<"ResultOutlet::run. closure="<<closure<<endl;
  jframe->x = closure->resultInt();
}



/*----------------------GFrame------------------*/	

void GFrame::setOutletOn(Closure *c) { /*assert(0); abort(); */}
Closure *GFrame::makeClosure() { /*assert(0); abort(); */return NULL;}
GFrame::GFrame():JobFrame(), PC(0) { }
GFrame *GFrame::copy() { return new GFrame(*this); }

/*---------------------Job---------------------------*/

// Job::Job() {}
// Job::Job(Pool *pool) : Job::Job(new JobFrame(), pool) {}

Job::Job(Pool *pool, Frame *f) : Closure(f), jobDone(false) {
		this->pool=pool;
		parent = NULL;
		joinCount=0;
		status = READY;
		pthread_cond_init (&cond_done, NULL);
}
Job::~Job() {
  pthread_cond_destroy(&cond_done);
}
void Job::compute(Worker *w, Frame *frame) {
		int x;
		JobFrame *f = (JobFrame *) frame;
		switch (f->PC) {
		case LABEL_0: 
			f->PC=LABEL_1;
			MEM_BARRIER(); // TODO RAJ
			// spawning
			x = spawnTask(w);
			if(w->abortOnSteal(x)) {
			  return ;
			}
			f->x=x;
		case LABEL_1: 
			f->PC=LABEL_2;
			MEM_BARRIER(); // TODO RAJ
			if (sync(w)) return;
		case LABEL_2: 
			setResultInt(f->x);
			setupReturn(w);
		default: return;
		}
		return;
}
int Job::spawnTask(Worker *ws) { assert(0); abort(); return 0; } // TODO RAJ child must provide an imple for this
void Job::completed() {
	
	Closure::completed();
	pthread_cond_broadcast(&cond_done);
	pool->jobCompleted();
}
void Job::waitForCompletion() {
		while (!isDone()) { pthread_cond_wait(&cond_done,NULL); }
}
    
bool Job::isCancelled() const { return false;}
    
bool Job::cancel(bool b) const { return false;}

int Job::getInt() {
//#warning "Need to check sched_yield() and locks, etc."
//   while(!isDone()) 
//     sched_yield();
  while(!isJobDone()) 
    sched_yield();
  return resultInt();
}

void Job::jobCompleted() {
  jobDone = true;
}

bool Job::isJobDone() volatile { return jobDone; }

/*--------------------GloballyQuiescentJob-------------------*/

bool GloballyQuiescentJob::requiresGlobalQuiescence() { return true;}	
// GloballyQuiescentJob::GloballyQuiescentJob(Pool *pool) 
//   : GloballyQuiescentJob(pool, new GFrame()) {}

GloballyQuiescentJob::GloballyQuiescentJob(Pool *pool, Frame *f) 
  : Job(pool, f) {
  parent = NULL;
  joinCount=0;
  status = READY;
}

void GloballyQuiescentJob::compute(Worker *w, Frame *frame) {
			GFrame *f = (GFrame *) frame;
			int PC = f->PC;
			f->PC=LABEL_1;
			MEM_BARRIER(); // TODO Raj
			if (PC==0) {
				// spawning
				int x = spawnTask(w);
				if(w->abortOnSteal(x))
				  return ;
				f->x=x;
				// Accumulate into result.
				int old = resultInt();
				accumulateResultInt(f->x);
				//if (Worker.reporting)
				//System.out.println( w + " " + this + " adds " + f.x + " to move " + old + " --> " + resultInt());
			}
			setupGQReturn(w);
}



/*------------------------GloballyQuiescentVoidJob----------------*/

/*virtual*/ void 
GloballyQuiescentVoidJob::compute(Worker *w, Frame *frame) { 
  assert(!w->hasThrownException());
  frame->compute(w);

  if(w->cache->interrupted()) {
    w->throwException();
    return;
  }
  //w->cache->reset();
  // The completion of the job might leave behind work (frames).
  // Do not pop the framestack.
  setupGQReturnNoArgNoPop(w);
  //setupGQReturnNoArg(w);
}
